src/lib.rs

```diff
@@ -21,14 +21,41 @@ pub(crate) fn app_dir(app_file: impl AsRef<Path>) -> Result<PathBuf> {
     Ok(path_buf)
 }
 
-pub(crate) fn write_failed_msg(bindle_id: &bindle::Id, dest_dir: &Path) -> String {
+pub(crate) fn push_all_failed_msg(path: &Path, server_url: &str) -> String {
     format!(
-        "Failed to write bindle '{}' to {}",
-        bindle_id,
-        dest_dir.display()
+        "Failed to push bindle from '{}' to the server at '{}'",
+        path.display(),
+        server_url
     )
 }
 
+pub(crate) fn missing_build_artifact_msg(path: impl AsRef<Path>) -> String {
+    format!(
+        "Component build artifact is missing: {}\n\nPlease try to run `spin build` first",
+        path.as_ref().display()
+    )
+}
+
+pub(crate) async fn prepare_bindle(
+    app_file: impl AsRef<Path>,
+    buildinfo: Option<semver::BuildMetadata>,
+    dest_dir: impl AsRef<Path>,
+) -> Result<bindle::Id> {
+    let result = spin_publish::prepare_bindle(&app_file, buildinfo, &dest_dir).await;
+
+    match result {
+        Err(spin_publish::Error::BindleComponentManifest {
+            source: source_err,
+            path,
+        }) if source_err.kind() == std::io::ErrorKind::NotFound => {
```

Comment:
Could we detect this in `spin_publish::prepare_bindle` and make the error more specific?  E.g. `Err(PublishError::MissingArtifact(path)) => ...`.  That would simplify the pattern match - currently the multiple lines make the flow hard for me to follow (though this is partly down to `rustfmt`!)

Reply:
done